### PR TITLE
Maybe utilities: `find`, `head`, `last`, `all`, and `tuple`!

### LINKED ISSUE #19: More Maybe helpers (for arrays)
- [ ] `Maybe.find`
- [ ] `Maybe.head`

### PR DESCRIPTION
**Pull request by chriskrycho**
There are some weird shenanigans in here type-wise—places where I can write the type with conditional types but TS can't figure out that what I'm returning actually matches that. Resolves #19.

### CODE REVIEW

File: src/maybe.ts [hunk 1]
```diff
@@ -1270,18 +1272,217 @@ export function isInstance<T = any>(item: any): item is Maybe<T> {
...
+  homogeneous arrays when using that approach.
+
+  @param args The `Maybe`s to resolve to a single `Maybe`.
+ */
+function all<T extends Maybe<any>>(...args: T[]): T extends Maybe<infer U> ? Maybe<U[]> : never {
+  // @ts-ignore -- this is indeed the correct implementation, but TS doesn't
+  //               correctly parse the types in the context of `reduce`.
+  return args.reduce(
+    (result, maybe) => result.andThen(as => maybe.map(a => as.concat(a))),
+    Maybe.just([] as any[])
```
Comment (bmakuh) [inline src/maybe.ts]: Why not `Maybe.just([] as T[])`?

File: src/maybe.ts [hunk 2]
```diff
@@ -1270,18 +1272,217 @@ export function isInstance<T = any>(item: any): item is Maybe<T> {
...
+
+  type Tuple = [Maybe<string>, Maybe<number>];
+
+  let invalid: Tuple = [Maybe.just('wat'), Maybe.nothing()];
+  let result = Maybe.tuple(invalid);
+  ```
+
+  If all of the items in the tuple are `Just`, the result is `Just` wrapping the
+  tuple of the values of the items. Here, for example, `result` again has the
+  type `Maybe<[string, number]>` and will be `Just['hey', 12]`:
```
Comment (bmakuh) [inline src/maybe.ts]: `Just(['hey', 12])`

File: src/maybe.ts [hunk 3]
```diff
@@ -1270,18 +1272,217 @@ export function isInstance<T = any>(item: any): item is Maybe<T> {
...
+
+  ```ts
+  import Maybe from 'true-myth/maybe';
+
+  type Tuple = [Maybe<string>, Maybe<number>];
+
+  let valid: Tuple = [Maybe.just('hey'), Maybe.just(12)];
+  let result = Maybe.tuple(valid);
```
Comment (bmakuh) [inline src/maybe.ts]: `let result = Maybe.tuple(valid); // => Just(['hey', 12])`

File: src/maybe.ts (hunk 1 above)
Comment (chriskrycho) [reply src/maybe.ts]: I tried it; it doesn't type-check either. 🤕 

When TS 3.1 lands, I'll be able to switch this up nicely, but in the meantime it's this gross hack.

Comment (chriskrycho) [reply src/maybe.ts]: …weirdly, I just tried this and it worked, but I could swear I've tried it before and had it fail. ¯\\\_(ツ)_/¯

**Pull request merged**